discovery-provider/src/tasks/index.py

```diff
@@ -363,306 +504,106 @@ def index_blocks(self, db, blocks_list):
-                    == get_contract_addresses()["user_factory"]
-                ):
-                    logger.info(
-                        f"index.py | UserFactory contract addr: {tx_target_contract_address}"
-                        f" tx from block - {tx}, receipt - {tx_receipt}, adding to user_factory_txs to process in bulk"
+            skip_whole_block = skip_tx_hash == "commit"  # db tx failed at commit level
+            if skip_whole_block:
+                logger.info(f"index.py | Skipping all txs in block {block.hash}")
+            else:
+                add_indexed_block_to_db(session, block)
+                TXS_GROUPED_BY_TYPE = {
+                    USER_FACTORY: [],
+                    TRACK_FACTORY: [],
+                    SOCIAL_FEATURE_FACTORY: [],
+                    PLAYLIST_FACTORY: [],
+                    USER_LIBRARY_FACTORY: [],
+                    USER_REPLICA_SET_MANAGER: [],
+                }
+
+                tx_receipt_dict = fetch_tx_receipts(self, block.transactions)
+
+                # Sort transactions by hash
+                sorted_txs = sorted(block.transactions, key=lambda entry: entry["hash"])
+
+                # Parse tx events in each block
+                for tx in sorted_txs:
+                    tx_hash = web3.toHex(tx["hash"])
+                    tx_target_contract_address = tx["to"]
+                    tx_receipt = tx_receipt_dict[tx_hash]
+                    should_skip_tx = (tx_target_contract_address == zero_address) or (
+                        skip_tx_hash is not None and skip_tx_hash == tx_hash
                     )
-                    user_factory_txs.append(tx_receipt)
 
-                # Handle track operations
-                if (
-                    tx_target_contract_address
-                    == get_contract_addresses()["track_factory"]
-                ):
-                    logger.info(
-                        f"index.py | TrackFactory contract addr: {tx_target_contract_address}"
-                        f" tx from block - {tx}, receipt - {tx_receipt}"
-                    )
-                    # Track state operations
-                    track_factory_txs.append(tx_receipt)
-
-                # Handle social operations
-                if (
-                    tx_target_contract_address
-                    == get_contract_addresses()["social_feature_factory"]
-                ):
-                    logger.info(
-                        f"index.py | Social feature contract addr: {tx_target_contract_address}"
-                        f"tx from block - {tx}, receipt - {tx_receipt}"
+                    if should_skip_tx:
+                        logger.info(
+                            f"index.py | Skipping tx {tx_hash} targeting {tx_target_contract_address}"
+                        )
+                        continue
+                    else:
+                        group_transaction_by_type(TXS_GROUPED_BY_TYPE, tx, tx_receipt)
```

Comment:
maybe this would be clearer if `group_transactions_by_type` returned a map instead of modifying an existing one?

Reply:
+1 to no mutation